### PR TITLE
Add type inference for constants

### PR DESCRIPTION
**Pull request by kehemo**
not working yet, need major changes

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -26,3 +26,4 @@ configure
 dependencies/chipyard
 
 .vscode
+scratch/
```
Comment (yamaguchi1024) [inline .gitignore:29]: remove this line

File: tests/golden/test_window/test_window_stmt.txt [hunk 2]
```diff
@@ -48,7 +48,7 @@ void window_stmt( void *ctxt, int_fast32_t n, int_fast32_t m, const float* x );
 void window_stmt( void *ctxt, int_fast32_t n, int_fast32_t m, const float* x ) {
 struct exo_win_1f32c y = (struct exo_win_1f32c){ &x[0], { m } };
 float *z = (float*) malloc(n * sizeof(*z));
-for (int_fast32_t i = 0; i < n; i++) {
+for (int_fast32_t i = ((int_fast32_t) 0); i < n; i++) {
```
Comment (yamaguchi1024) [inline tests/golden/test_window/test_window_stmt.txt:51]: Remove the casting for index expressions

File: tests/test_codegen.py [hunk 3]
```diff
@@ -530,3 +530,18 @@ def foo(x: i8[10]):
...
+    def foo(x: f64[1], y: f32[1]):
+        x[0] = (16777216.0 + 1.0) + y[0]
+
+    # all consts should coerce to f32
+    # so, 16777216 + 1 should yield 16777216 again
+    x = np.array([3.0], dtype=np.float64)
+    y = np.array([1.0], dtype=np.float32)
+    fn = compiler.compile(foo)
+    fn(None, x, y)
+    assert x[0] == 16777216
```
Comment (yamaguchi1024) [inline tests/test_codegen.py:547]: We need individual tests for different buffer types, e.g., 
a : i8
a + 3

**Review by yamaguchi1024 — CHANGES_REQUESTED** [review]

File: src/exo/typecheck.py [hunk 4]
```diff
@@ -438,44 +439,42 @@ def check_e(self, e):
             typ = T.err
             if lhs.type == T.err or rhs.type == T.err:
                 typ = T.err
-            elif e.op == "and" or e.op == "or":
```
Comment (yamaguchi1024) [inline src/exo/typecheck.py:441]: Change check_e to `check_e(e, is_index=False)`

File: src/exo/typecheck.py [hunk 5]
```diff
@@ -438,44 +439,42 @@ def check_e(self, e):
...
-                if lhs.type is not T.bool:
-                    self.err(lhs, "expected 'bool' argument to logical op")
```
Comment (yamaguchi1024) [inline src/exo/typecheck.py:443]:
```
if is_index:
   ty = {float: T.R, bool: T.bool, int: T.int}.get(type(e.val))
else:
   ty = {float: T.R, bool: T.bool, int: T.R}.get(type(e.val))
```

File: src/exo/LoopIR_compiler.py [hunk 6]
```diff
@@ -994,7 +994,7 @@ def comp_e(self, e, prec=0):
         elif isinstance(e, LoopIR.Const):
             if isinstance(e.val, bool):
                 return "true" if e.val else "false"
-            return str(e.val)
+            return f"(({e.type.ctype()}) {str(e.val)})"
```
Comment (yamaguchi1024) [inline src/exo/LoopIR_compiler.py:997]:
```
if e.is_indexable():
    return str(e.val)
else:
   if e.base_typ() == T.f32:
     return str(e.val)+'f'
   if e.base_typ() == T.f64 or e.base_typ() == T.i32:
      return str(e.val)
   return f"(({e.type.ctype()}) {str(e.val)})"
```

**Comment by kehemo** [comment]
BLAS macos-14 testing check appeared to fail first time due to nondeterministic error in formatter
[https://github.com/google/yapf/issues/1164](https://github.com/google/yapf/issues/1164)

File: src/exo/LoopIR_compiler.py [hunk 7]
```diff
@@ -994,7 +994,14 @@ def comp_e(self, e, prec=0):
         elif isinstance(e, LoopIR.Const):
             if isinstance(e.val, bool):
                 return "true" if e.val else "false"
-            return str(e.val)
+            elif e.type.is_indexable():
+                return f"{int(e.val)}"
+            elif e.type == T.f64:
+                return f"{float(e.val)}"
```
Comment (yamaguchi1024) [inline src/exo/LoopIR_compiler.py:1000]: Why do you cast to float when T.f64? It could underflow, right?

Comment (yamaguchi1024) [reply src/exo/LoopIR_compiler.py:1000]: Oh it's Python float, nvm

**Review by yamaguchi1024 — APPROVED** [review]
Looks good to me! Please write a commit message when you merge

**Pull request merged**